README.md
# MDMDirector

MDMDirector is an opinionated orchestrator for [MicroMDM](https://github.com/micromdm/micromdm). It enables profiles to be managed with MicroMDM in a stateful manner, via a RESTful API. It also allows for installation of packages either just at enrollment or immediately. It receives webhook events from MicroMDM and then instructs MicroMDM to perform appropriate actions. As such, MDMDirector does not need to be exposed to the public internet.

## Usage

MDMDirector is a compiled binary and is configured using flags.

Requirements:

* Redis for the scheduled checkin queue
* PostgreSQL database for storing device information
* (Recommended) Signing certificate for signing profiles
* (STRONGLY recommended) Load balancer/proxy to serve and terminate TLS for MDMDirector


### MicroMDM Setup

You must set the `-command-webhook-url` flag on MicroMDM to the URL of your MDMDirector instance (with the addition of `/webhook`).

```
-command-webhook-url=https://mdmdirector.company.com/webhook
```

### Flags

- `-cert /path/to/certificate` - Path to the signing certificate or p12 file.
- `-clear-device-on-enroll` - Deletes device profiles and install applications when a device enrolls (default "false")
- `-db-host string` - **(Required)** Hostname or IP of the PostgreSQL instance
- `-db-max-connections int` - Maximum number of database connections (default 100)
- `-db-name string` - **(Required)** Name of the database to connect to.
- `-db-password string` - **(Required)** Password of the DB user.
- `-db-port string` - The port of the PostgreSQL instance (default 5432)
- `-db-sslmode` - The SSL Mode to use to connect to PostgreSQL (default "disable")
- `-db-username string` - **(Required)** Username used to connect to the PostgreSQL instance.
- `-redis-host string` - Hostname of your Redis instance (default "localhost").
- `-redis-port string` - Port of your Redis instance (default 6379).
- `-redis-password string` - Password for your Redis instance (default is no password).
- `-debug` - Enable debug mode. Does things like shorten intervals for scheduled tasks. Only to be used during development.
- `-enrollment-profile` - Path to enrollment profile.
- `-enrollment-profile-signed` - Is the enrollment profile you are providing already signed (default: false)
- `-escrowurl` - HTTP(S) endpoint to escrow erase and unlock PINs to ([Crypt](https://github.com/grahamgilbert/crypt-server) and other compatible servers).
- `info-request-interval` - The amount of time in minutes to wait before requesting `DeviceInfo`, `ProfileList`, `SecurityInfo` etc. Defaults to 360.
- `-key-password string` - Password to decrypt the signing key or p12 file.
- `-loglevel string` - Log level. One of debug, info, warn, error (default "warn")
- `-logformat-format` - Log format. Either `logfmt` (the default) or `json`.
- `-micromdmapikey string` - **(Required)** MicroMDM Server API Key.
- `-micromdmurl string` - **(Required)** MicroMDM Server URL.
- `-once-in` - Number of minutes to wait before queuing an additional command for any device which already has commands queued. Defaults to 60. Ignored and overridden as 2 (minutes) if --debug is passed.
- `-password string` - **(Required)** Password used for basic authentication
- `-port string` - Port number to run MDMDirector on. (default "8000")
- `-prometheus` - Enable Prometheus metrics. (default false)
- `-push-new-build` - Re-push profiles if the device's build number changes. (default true)
- `-scep-cert-issuer` - The issuer of your SCEP certificate (default: "CN=MicroMDM,OU=MICROMDM SCEP CA,O=MicroMDM,C=US")
- `-scep-cert-min-validity` - The number of days at which the SCEP certificate has remaining before the enrollment profile is re-sent. (default: 180)
- `-sign` - Sign profiles prior to sending to MicroMDM. Requires `-cert` to be passed.
- `-signing-private-key string` - Path to the signing private key. Don't use with p12 file.


## [Todo](/mdmdirector/getting_involved_docs/start_here.md)

## [Contributing](/mdmdirector/getting_involved_docs/start_here.md)



CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v0.5.2"></a>
## v0.5.2 - 2022-10-23
### Pull Requests
- Merge pull request [#82](https://github.com/bizzybits/mdmdirector/issues/82) from mdmdirector/verify_profile_cert
- Merge pull request [#80](https://github.com/bizzybits/mdmdirector/issues/80) from mdmdirector/vars
- Merge pull request [#81](https://github.com/bizzybits/mdmdirector/issues/81) from mdmdirector/indexing
- Merge pull request [#76](https://github.com/bizzybits/mdmdirector/issues/76) from williamtheaker/patch-3
- Merge pull request [#75](https://github.com/bizzybits/mdmdirector/issues/75) from daemonsy/example_typo
- Merge pull request [#74](https://github.com/bizzybits/mdmdirector/issues/74) from mdmdirector/shared_profiles_comparison
- Merge pull request [#73](https://github.com/bizzybits/mdmdirector/issues/73) from discentem/log_profileList_data
- Merge pull request [#72](https://github.com/bizzybits/mdmdirector/issues/72) from mdmdirector/verifymdmprofiles_refactor
- Merge pull request [#71](https://github.com/bizzybits/mdmdirector/issues/71) from discentem/log_profileList_contente
- Merge pull request [#69](https://github.com/bizzybits/mdmdirector/issues/69) from mdmdirector/fix_cleanup
- Merge pull request [#70](https://github.com/bizzybits/mdmdirector/issues/70) from discentem/log_profileList_contente
- Merge pull request [#67](https://github.com/bizzybits/mdmdirector/issues/67) from mdmdirector/move_cleanup_function
- Merge pull request [#66](https://github.com/bizzybits/mdmdirector/issues/66) from discentem/logging_fix_verifyprofiles
- Merge pull request [#65](https://github.com/bizzybits/mdmdirector/issues/65) from discentem/for_loop_syntax
- Merge pull request [#64](https://github.com/bizzybits/mdmdirector/issues/64) from mdmdirector/stop_deleting
- Merge pull request [#63](https://github.com/bizzybits/mdmdirector/issues/63) from mdmdirector/fix_duplicate_profiles
- Merge pull request [#61](https://github.com/bizzybits/mdmdirector/issues/61) from mdmdirector/push_length
- Merge pull request [#60](https://github.com/bizzybits/mdmdirector/issues/60) from discentem/patch-1
- Merge pull request [#58](https://github.com/bizzybits/mdmdirector/issues/58) from mdmdirector/push_endpoint
- Merge pull request [#59](https://github.com/bizzybits/mdmdirector/issues/59) from discentem/onceIn_readme
- Merge pull request [#57](https://github.com/bizzybits/mdmdirector/issues/57) from discentem/configurable_OnceIn
- Merge pull request [#56](https://github.com/bizzybits/mdmdirector/issues/56) from discentem/fix_logging_chunk_progress
- Merge pull request [#55](https://github.com/bizzybits/mdmdirector/issues/55) from mdmdirector/one_push_at_a_time
- Merge pull request [#54](https://github.com/bizzybits/mdmdirector/issues/54) from discentem/spread-checkin-load
- Merge pull request [#52](https://github.com/bizzybits/mdmdirector/issues/52) from mdmdirector/redis
- Merge pull request [#53](https://github.com/bizzybits/mdmdirector/issues/53) from mdmdirector/revert_profiles
- Merge pull request [#51](https://github.com/bizzybits/mdmdirector/issues/51) from mdmdirector/more_logs
- Merge pull request [#50](https://github.com/bizzybits/mdmdirector/issues/50) from mdmdirector/request_info_now
- Merge pull request [#48](https://github.com/bizzybits/mdmdirector/issues/48) from mdmdirector/calm_down
- Merge pull request [#49](https://github.com/bizzybits/mdmdirector/issues/49) from mdmdirector/chill_out_profiles
- Merge pull request [#47](https://github.com/bizzybits/mdmdirector/issues/47) from mdmdirector/prometheus_flag
- Merge pull request [#46](https://github.com/bizzybits/mdmdirector/issues/46) from mdmdirector/fix_verifymdmprofiles
- Merge pull request [#45](https://github.com/bizzybits/mdmdirector/issues/45) from mdmdirector/fix_security_info
- Merge pull request [#42](https://github.com/bizzybits/mdmdirector/issues/42) from Ryxias/dw--unit-tests
- Merge pull request [#44](https://github.com/bizzybits/mdmdirector/issues/44) from williamtheaker/patch-2
- Merge pull request [#43](https://github.com/bizzybits/mdmdirector/issues/43) from mdmdirector/send_commands
- Merge pull request [#41](https://github.com/bizzybits/mdmdirector/issues/41) from mdmdirector/gorm2
- Merge pull request [#40](https://github.com/bizzybits/mdmdirector/issues/40) from mdmdirector/fix_command_loop
- Merge pull request [#39](https://github.com/bizzybits/mdmdirector/issues/39) from mdmdirector/handle_zero_values
- Merge pull request [#37](https://github.com/bizzybits/mdmdirector/issues/37) from mdmdirector/fix_profile_list_comparison
- Merge pull request [#36](https://github.com/bizzybits/mdmdirector/issues/36) from mdmdirector/update_securityinfo
- Merge pull request [#38](https://github.com/bizzybits/mdmdirector/issues/38) from mdmdirector/individual_requests
- Merge pull request [#35](https://github.com/bizzybits/mdmdirector/issues/35) from mdmdirector/fix_disabling_shared
- Merge pull request [#34](https://github.com/bizzybits/mdmdirector/issues/34) from mdmdirector/no_escrow_url
- Merge pull request [#33](https://github.com/bizzybits/mdmdirector/issues/33) from williamtheaker/patch-1
- Merge pull request [#31](https://github.com/bizzybits/mdmdirector/issues/31) from gavinelder/tools
- Merge pull request [#30](https://github.com/bizzybits/mdmdirector/issues/30) from mdmdirector/scep_renewal
- Merge pull request [#29](https://github.com/bizzybits/mdmdirector/issues/29) from mdmdirector/logging
- Merge pull request [#28](https://github.com/bizzybits/mdmdirector/issues/28) from mdmdirector/update_stale_info
- Merge pull request [#26](https://github.com/bizzybits/mdmdirector/issues/26) from mdmdirector/prometheus
- Merge pull request [#25](https://github.com/bizzybits/mdmdirector/issues/25) from mdmdirector/uuid
- Merge pull request [#24](https://github.com/bizzybits/mdmdirector/issues/24) from mdmdirector/erase_lock

main.go
package main

import (
	"flag"
	"net/http"
	"time"

	"github.com/gorilla/mux"
	"github.com/mdmdirector/mdmdirector/db"
	"github.com/mdmdirector/mdmdirector/director"
	"github.com/mdmdirector/mdmdirector/types"
	"github.com/mdmdirector/mdmdirector/utils"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/vmihailenco/taskq/v3"
	"github.com/vmihailenco/taskq/v3/redisq"

	"github.com/micromdm/go4/env"
	log "github.com/sirupsen/logrus"
)

// MicroMDMURL is the url for your MicroMDM server
var MicroMDMURL string

// MicroMDMAPIKey is the api key for your MicroMDM server
var MicroMDMAPIKey string

// Sign is whether profiles should be signed (they really should)
var Sign bool

// KeyPassword is the password for the private key
var KeyPassword string

// KeyPath is the path for the private key
var KeyPath string

// CertPath is the path for the signing cert or p12 file
var CertPath string

// PushNewBuild is whether to push all profiles if the device's build number changes
var PushNewBuild bool

// BasicAuthPass is the password used for basic auth
var BasicAuthPass string

// DBUsername is the account to connect to the database
var DBUsername string

// DBPassword is used to connect to the database
var DBPassword string

// DBName is used to connect to the database
var DBName string

// DBHost is used to connect to the database
var DBHost string

// DBPort is used to connect to the database
var DBPort string

var DBMaxConnections int

// DBSSLMode is used to connect to the database
var DBSSLMode string

// LogLevel = log level
var LogLevel string

// EscrowURL = url to escrow erase and lock PINs to
var EscrowURL string

var ClearDeviceOnEnroll bool

var ScepCertIssuer string

var ScepCertMinValidity int

var EnrollmentProfile string

var SignEnrollmentProfile bool

var LogFormat string

var Prometheus bool

var RedisHost string

var RedisPort string

var RedisPassword string

var OnceIn int

var InfoRequestInterval int

func main() {
	var port string
	var debugMode bool
	flag.BoolVar(&debugMode, "debug", env.Bool("DEBUG", false), "Enable debug mode")
	flag.BoolVar(&PushNewBuild, "push-new-build", env.Bool("PUSH_NEW_BUILD", true), "Re-push profiles if the device's build number changes.")
	flag.StringVar(&port, "port", env.String("DIRECTOR_PORT", "8000"), "Port number to run mdmdirector on.")
	flag.StringVar(&MicroMDMURL, "micromdmurl", env.String("MICRO_URL", ""), "MicroMDM Server URL")
	flag.StringVar(&MicroMDMAPIKey, "micromdmapikey", env.String("MICRO_API_KEY", ""), "MicroMDM Server API Key")
	flag.BoolVar(&Sign, "sign", env.Bool("SIGN", false), "Sign profiles prior to sending to MicroMDM.")
	flag.StringVar(&KeyPassword, "key-password", env.String("SIGNING_PASSWORD", ""), "Password to encrypt/read the signing key(optional) or p12 file.")
	flag.StringVar(&KeyPath, "signing-private-key", env.String("SIGNING_KEY", ""), "Path to the signing private key. Don't use with p12 file.")
	flag.StringVar(&CertPath, "cert", env.String("SIGNING_CERT", ""), "Path to the signing certificate or p12 file.")
	flag.StringVar(&BasicAuthPass, "password", env.String("DIRECTOR_PASSWORD", ""), "Password used for basic authentication")
	flag.StringVar(&DBUsername, "db-username", env.String("DB_USERNAME", ""), "The username associated with the Postgres instance")
	flag.StringVar(&DBPassword, "db-password", env.String("DB_PASSWORD", ""), "The password of the db user account")
	flag.StringVar(&DBName, "db-name", env.String("DB_NAME", ""), "The name of the Postgres database to use")
	flag.StringVar(&DBHost, "db-host", env.String("DB_HOST", ""), "The hostname or IP of the Postgres instance")
	flag.StringVar(&DBPort, "db-port", env.String("DB_PORT", "5432"), "The port of the Postgres instance")
	flag.StringVar(&RedisHost, "redis-host", env.String("REDIS_HOST", "localhost"), "Redis hostname")
	flag.StringVar(&RedisPort, "redis-port", env.String("REDIS_PORT", "6379"), "Redis port")
	flag.StringVar(&RedisPassword, "redis-password", env.String("REDIS_PASSWORD", ""), "Redis password")
	flag.StringVar(&DBSSLMode, "db-sslmode", "disable", "The SSL Mode to use to connect to Postgres")
	flag.IntVar(&DBMaxConnections, "db-max-connections", 100, "Maximum number of database connections")
	flag.StringVar(&LogLevel, "loglevel", env.String("LOG_LEVEL", "warn"), "Log level. One of debug, info, warn, error")
	flag.StringVar(&EscrowURL, "escrowurl", env.String("ESCROW_URL", ""), "HTTP endpoint to escrow erase and unlock PINs to.")
	flag.BoolVar(&ClearDeviceOnEnroll, "clear-device-on-enroll", env.Bool("CLEAR_DEVICE_ON_ENROLL", false), "Deletes device profiles and install applications when a device enrolls")
	flag.StringVar(&LogFormat, "log-format", env.String("LOG_FORMAT", "logfmt"), "Format to output logs. Defaults to logfmt. Can be set to logfmt or json.")
	flag.StringVar(&ScepCertIssuer, "scep-cert-issuer", env.String("SCEP_CERT_ISSUER", "OU=MICROMDM SCEP CA,O=MicroMDM,C=US"), "The issuer of your SCEP certificate")
	flag.IntVar(&ScepCertMinValidity, "scep-cert-min-validity", env.Int("SCEP_CERT_MIN_VALIDITY", 180), "The number of days at which the SCEP certificate has remaining before the enrollment profile is re-sent.")
	flag.StringVar(&EnrollmentProfile, "enrollment-profile", env.String("ENROLLMENT_PROFILE", ""), "Path to enrollment profile.")
	flag.BoolVar(&SignEnrollmentProfile, "enrollment-profile-signed", env.Bool("ENROLMENT_PROFILE_SIGNED", false), "Is the enrollment profile you are providing already signed")
	flag.BoolVar(&Prometheus, "prometheus", env.Bool("PROMETHEUS", false), "Enable Prometheus")
	flag.IntVar(&OnceIn, "once-in", env.Int("ONCE_IN", 60), "Number of minutes to wait before queuing an additional command for any device which already has commands queued. Defaults to 60. Ignored and overidden as 2 (minutes) if --debug is passed.")
	flag.IntVar(&InfoRequestInterval, "info-request-interval", env.Int("INFO_REQUEST_INTERVAL", 360), "Number of minutes to wait between issuing information commands")
	flag.Parse()

	logLevel, err := log.ParseLevel(LogLevel)
	if err != nil {
		log.Fatalf("Unable to parse the log level - %s \n", err)
	}
	log.SetLevel(logLevel)

	if LogFormat == "json" {
		log.SetFormatter(&log.JSONFormatter{
			DisableHTMLEscape: true,
		})
	} else {
		log.SetFormatter(&log.TextFormatter{
			FullTimestamp: true,
			DisableQuote:  true,
		})
	}

	if MicroMDMURL == "" {
		log.Fatal("MicroMDM Server URL missing. Exiting.")
	}

	if MicroMDMAPIKey == "" {
		log.Fatal("MicroMDM API Key missing. Exiting.")
	}

	if BasicAuthPass == "" {
		log.Fatal("Basic Auth password missing. Exiting.")
	}

	if DBUsername == "" || DBPassword == "" || DBName == "" || DBHost == "" || DBPort == "" || DBSSLMode == "" {
		log.Fatal("Required database details missing, Exiting.")
	}

	if LogLevel != "debug" && LogLevel != "info" && LogLevel != "warn" && LogLevel != "error" {
		log.Fatal("loglevel value is not one of debug, info, warn or error.")
	}

	r := mux.NewRouter()
	r.HandleFunc("/webhook", director.WebhookHandler).Methods("POST")
	r.HandleFunc("/profile", utils.BasicAuth(director.PostProfileHandler)).Methods("POST")
	r.HandleFunc("/profile", utils.BasicAuth(director.DeleteProfileHandler)).Methods("DELETE")
	r.HandleFunc("/profile", utils.BasicAuth(director.GetSharedProfiles)).Methods("GET")
	r.HandleFunc("/profile/{udid}", utils.BasicAuth(director.GetDeviceProfiles)).Methods("GET")
	r.HandleFunc("/device", utils.BasicAuth(director.DeviceHandler)).Methods("GET")
	r.HandleFunc("/device/command/{command}", utils.BasicAuth(director.PostDeviceCommandHandler)).Methods("POST")
	r.HandleFunc("/device/serial/{serial}", utils.BasicAuth(director.SingleDeviceSerialHandler)).Methods("GET")
	r.HandleFunc("/device/push/{udid}", utils.BasicAuth(director.PushDeviceHandler)).Methods("GET")
	r.HandleFunc("/device/{udid}", utils.BasicAuth(director.SingleDeviceHandler)).Methods("GET")
	r.HandleFunc("/installapplication", utils.BasicAuth(director.PostInstallApplicationHandler)).Methods("POST")
	r.HandleFunc("/installapplication", utils.BasicAuth(director.GetSharedApplicationss)).Methods("GET")
	r.HandleFunc("/command/pending", utils.BasicAuth(director.GetPendingCommands)).Methods("GET")
	r.HandleFunc("/command/pending/delete", utils.BasicAuth(director.DeletePendingCommands)).Methods("GET")
	r.HandleFunc("/command/error", utils.BasicAuth(director.GetErrorCommands)).Methods("GET")
	r.HandleFunc("/command", utils.BasicAuth(director.GetAllCommands)).Methods("GET")
	r.HandleFunc("/health", director.HealthCheck).Methods("GET")

	director.InfoLogger(director.LogHolder{Message: "Connecting to database"})
	if err := db.Open(); err != nil {
		director.ErrorLogger(director.LogHolder{Message: err.Error()})
		log.Fatal("Failed to open database")
	}
	director.InfoLogger(director.LogHolder{Message: "Connected to database"})

	director.InfoLogger(director.LogHolder{Message: "Performing DB migrations if required"})

	err = db.DB.AutoMigrate(
		&types.Device{},
		&types.DeviceProfile{},
		&types.Command{},
		&types.SharedProfile{},
		&types.SecurityInfo{},
		&types.FirmwarePasswordStatus{},
		&types.ManagementStatus{},
		&types.OSUpdateSettings{},
		&types.FirewallSettings{},
		// &types.FirewallSettingsApplication{},
		&types.SecureBoot{},
		&types.SecureBootReducedSecurity{},
		&types.SharedInstallApplication{},
		&types.DeviceInstallApplication{},
		&types.Certificate{},
		&types.ProfileList{},
		&types.UnlockPin{},
	)
	if err != nil {
		director.ErrorLogger(director.LogHolder{Message: err.Error()})
		log.Fatal(err)
	}

	director.InfoLogger(director.LogHolder{Message: "mdmdirector is running, hold onto your butts..."})

	var QueueFactory = redisq.NewFactory()

	var PushQueue = QueueFactory.RegisterQueue(&taskq.QueueOptions{
		Name:  "pushnotifications",
		Redis: director.RedisClient(), // go-redis client
	})
	err = PushQueue.Purge()
	if err != nil {
		log.Error(err)
	}

	if utils.Prometheus() {
		director.Metrics()
		r.Handle("/metrics", promhttp.Handler())
	}

	go director.FetchDevicesFromMDM()

	// Override OnceIn if --debug is passed
	if debugMode {
		OnceIn = 2
	}

	onceInDuration := (time.Minute * time.Duration(OnceIn))
	go director.ScheduledCheckin(PushQueue, onceInDuration)
	go director.ProcessScheduledCheckinQueue(PushQueue)

	log.Info(http.ListenAndServe(":"+port, r))
}
